File: 01457. Pseudo-Palindromic Paths in a Binary Tree/01457.c
#define MAX_SIZE    10

void traversal(struct TreeNode *root, int map[], int odd_count, int *ans) {
    if (root != NULL) {
        ++(map[root->val]);
        odd_count += (map[root->val] % 2 == 1) ? 1 : -1;
        
        if (root->left != NULL || root->right != NULL) {
            traversal(root->left,  map, odd_count, ans);
            traversal(root->right, map, odd_count, ans);
        } else {
            if (odd_count <= 1) {
                ++(*ans);
            }
        }
        
        --(map[root->val]);
        // no need to do this, because the odd_count of parent is correct for itself
        // odd_count += (map[root->val] % 2 == 1) ? 1 : -1;
    }
}

int pseudoPalindromicPaths (struct TreeNode *root) {
    int ans = 0, map[MAX_SIZE] = {};
    
    traversal(root, map, 0, &ans);
    
    return ans;
}

File: 01457. Pseudo-Palindromic Paths in a Binary Tree/01457.go
// https://leetcode.com/problems/pseudo-palindromic-paths-in-a-binary-tree

func pseudoPalindromicPaths (root *TreeNode) int {
    return dfs(root, 0)
}

func dfs(node *TreeNode, bitmap int) (path int) {
    if node == nil {
        return 0
    }

    bitmap ^= (1 << node.Val)

    if node.Left == nil && node.Right == nil {
        // odd frequency in the bitmap, must be 0 or 1 count to form a Pseudo-Palindromic Paths
        if bitmap & (bitmap-1) == 0 {
            return 1
        }
        return 0
    }

    return dfs(node.Left, bitmap) + dfs(node.Right, bitmap)
}
